gapis/resolve/dependencygraph2/graph_visualization/graph_visualization.go

```diff
@@ -16,17 +16,208 @@ package graph_visualization
+		currentHierarchy.currentId++
+		currentHierarchy.currentLevel = currentLevel + 1
+	} else {
+		if currentLevel, ok := endCommands[commandName]; ok && currentLevel <= currentHierarchy.currentLevel {
+			currentHierarchy.currentLevel = currentLevel + 1
+			isEndCommand = true
+		}
+	}
+
+	label := "\""
+	for i := 0; i < currentHierarchy.currentLevel; i++ {
+		if i == 0 {
+			label += getCommandBuffer(command)
+		} else {
+			label += fmt.Sprintf("%s%d/", listOfCommandNames[i], currentHierarchy.levelId[i])
+		}
+	}
+	if isEndCommand {
+		currentHierarchy.currentLevel--
+	} else {
+		if _, ok := beginCommands[commandName]; ok {
+			if commandName == VK_CMD_BEGIN_RENDER_PASS {
+				currentHierarchy.levelId[currentHierarchy.currentLevel] = currentHierarchy.currentId
+				currentHierarchy.currentId++
+				currentHierarchy.currentLevel++
+			}
+		}
+	}
+	return label
+}
+
+func getSubCommandLabel(cmdNode dependencygraph2.CmdNode) string {
+	label := ""
+	for i := 1; i < len(cmdNode.Index); i++ {
+		label += fmt.Sprintf("/%d", cmdNode.Index[i])
+	}
+	return label
+}
 
+func getSplittedLabelByChar(label *string, splitChar byte) []string {
+	splitLabel := []string{}
+	lastPosition := 0
+	for i := 0; i <= len(*label); i++ {
+		if i == len(*label) || (*label)[i] == splitChar {
+			splitLabel = append(splitLabel, (*label)[lastPosition:i])
+			lastPosition = i + 1
+		}
+	}
+	return splitLabel
+}
+func getMaxCommonPrefixBetweenSplitLabels(splitLabel1 *[]string, splitLabel2 *[]string) int {
+	size := len(*splitLabel1)
+	if len(*splitLabel2) < size {
+		size = len(*splitLabel2)
+	}
+	for i := 0; i < size; i++ {
+		if (*splitLabel1)[i] != (*splitLabel2)[i] {
+			return i
+		}
+	}
+	return size
+}
+
+func getMaxCommonPrefixBetweenLabels(label1, label2 string) int {
+	splitLabel1 := getSplittedLabelByChar(&label1, '/')
+	splitLabel2 := getSplittedLabelByChar(&label2, '/')
+	return getMaxCommonPrefixBetweenSplitLabels(&splitLabel1, &splitLabel2)
+}
+
+func createGraphFromDependencyGraph(dependencyGraph dependencygraph2.DependencyGraph) (*Graph, error) {
+
+	numberOfNodes := dependencyGraph.NumNodes()
+	graph := createGraph(0)
+	currentHierarchy := &Hierarchy{}
+	previousNode := &Node{}
+	for i := 0; i < numberOfNodes; i++ {
+		dependencyNode := dependencyGraph.GetNode(dependencygraph2.NodeID(i))
+		if cmdNode, ok := dependencyNode.(dependencygraph2.CmdNode); ok {
+			cmdNodeId := cmdNode.Index[0]
+			command := dependencyGraph.GetCommand(api.CmdID(cmdNodeId))
+			commandName := command.CmdName()
+			label := getCommandLabel(currentHierarchy, command)
+			label += fmt.Sprintf("%s%d", commandName, cmdNodeId)
+			label += getSubCommandLabel(cmdNode)
+			label += "\""
+			attributes := fmt.Sprintf("\"%v\"", command.CmdParams())
+
+			graph.addNodeByIdAndLabelAndNameAndAttributes(i, label, commandName, attributes)
+
+			node := graph.nodeIdToNode[i]
+			if _, ok1 := commandsInsideRenderScope[previousNode.name]; ok1 {
```

Comment:
"Inside render scope" probably also needs to be abstracted in some way and moved into Vulkan-specific code.

Reply:
Yes, I moved this into Vulkan-specific code.